src/builtInPDBObjects/headers/Tree.h

```diff
@@ -376,9 +396,29 @@ namespace pdb
                         }
                     }
 
-                    outData[i] = (float)(tree[curIndex].returnClass);
+                    outData[rowNumBase+j] = (float)(tree[curIndex].returnClass);
+	        }
+            }
+            rowNumBase = numRows/batchSize*batchSize;
+	    base = rowNumBase * numCols;
+	    for (int i = 0; i < remainderSize; i++) {
+	    
+	       featureStartIndex = base + i*numCols;
+	       int curIndex = 0;
+	       while (tree[curIndex].isLeaf == false)
+               {
+                    if (inData[featureStartIndex + tree[curIndex].indexID] < tree[curIndex].returnClass)
+                    {
+                        curIndex = tree[curIndex].leftChild;
+                    }
+                    else
+                    {
+                        curIndex = tree[curIndex].rightChild;
+                    }
+               }
+	       outData[rowNumBase+i] = (float)(tree[curIndex].returnClass);
+	    
```

Comment:
Line 381 to line 420 can be reduced to 
```
        const bool hasRemainder = remainderSize != 0;
        const int iterations = numRows / batchSize + hasRemainder;
        // for (int i = 0; i < numRows / batchSize; i++) {
        for (int i = 0; i < iterations; i++) {
            base = i * batchSize * numCols;
            rowNumBase = i * batchSize;
            const int blockSize = (hasRemainder && i + 1 == iterations)
                                      ? remainderSize
                                      : batchSize;
            for (int j = 0; j < blockSize; j++) {
                featureStartIndex = base + j * numCols;
                int curIndex = 0;
                while (tree[curIndex].isLeaf == false) {
                    if (inData[featureStartIndex + tree[curIndex].indexID] <
                        tree[curIndex].returnClass) {
                        curIndex = tree[curIndex].leftChild;
                    } else {
                        curIndex = tree[curIndex].rightChild;
                    }
                }

                outData[rowNumBase + j] = (float)(tree[curIndex].returnClass);
            }
        }
```

Reply:
I didn't do that because I don't want the branch prediction in each outer loop, though the compiler and CPU branch prediction may do a good job.